### ISSUE TITLE
[C][Metachecker] Expression statement pattern with sizeof not matching expressions with sizeof

### ISSUE THREAD

**Issue by ben-elttam** [semgrep/semgrep#4038]
**Describe the bug**
Semgrep C matching is inconsistent in finding `sizeof`.

**To Reproduce**

* https://semgrep.dev/s/ben-elttam:c-sizeof  (finds all instances)
* https://semgrep.dev/s/ben-elttam:c-sizeof2 (doesn't find any instances)

I am running 0.68.2 locally and the behaviour is slighly different. I couldn't reproduce on semgrep.dev though.

Also changing the pattern from `sizeof($X)` to `sizeof(...)` gets a parse error, I believe this should be valid syntax though.

**Expected behavior**

All uses of `sizeof` should be found.
Also pattern `sizeof(...)` should work, similar to `sizeof($X)`.

I prefixed the lines with `// ruleid: ... ` syntax, you can see where it fails.

**Screenshots**

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [ ] P1: important to fix or quite annoying
- [X] P2: regular bug that should get fixed

At the moment I'm interested in C. I know support is alpha at the moment. I hope logging bugs helps mature support.

**Environment**
If not using semgrep.dev: are you running off docker, an official binary, a local build?

I typically use docker, but I also use official build via pip, and semgrep.dev when I can't get things to work how I want.

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/r2c/issue/PA-476/semgrep-c-inconsistent-matching-of-sizeof.

**Comment by aryx** [comment]
You had an extra semicolon in your pattern, so it will match only at the stmt level.
Try https://semgrep.dev/s/y13Z

**Comment by ben-elttam** [comment]
Thanks @aryx for pointing out the `;`.

On my localhost I didn't have a `;`. I had made a mistake when minimising the test case for the bug report.

Here is a more complete example, it should match two separate lines but it reports two matches both for line 18.
https://semgrep.dev/s/ben-elttam:c-sizeof3

Running this with returntocorp/semgrep:develop (Digest: sha256:475d6ae1f068f88c3e2ef8d9ea54b9af01ad258c75d5df00c115be48c3744714) I get corrected expected behaviour now.

Running it on my localhost I can see from the highlighted output that lines 11 and 18 are reported twice, once for the LHS `sizeof` and second time for the RHS `sizeof`.


----

NOTE: this was from before I realised semgrep:develop is working correctly. I've left it here, but its not important.

This may be related to the C parsing, as the C code is bad and generates compiler warnings. Here's clang warnings:

```
sizeof-array-param.c:11 [clang-tidy bugprone-sizeof-expression] Warning Suspicious usage of sizeof pointer 'sizeof(T*)/sizeof(T)'                                                                                                                                                                                           
sizeof-array-param.c:11 [clang -Wsizeof-array-argument] Warning Sizeof on array function parameter will return size of 'unsigned int *' instead of 'unsigned int [32]'
sizeof-array-param.c:11 [clang -Wsizeof-pointer-div] Warning 'sizeof (p)' will return the size of the pointer, not the array itself
```

Nevertheless, I want to be able to find such dodgy code with semgrep.
